### ISSUE TITLE
 Visualize inference results

### ISSUE THREAD

**Issue by qiqi-x** [kumuji/ugains#2]
Hello, thank you very much for your work. I would like to ask how to obtain the instance bounding boxes from model inference? Additionally, how can the inferred instance bounding boxes be displayed on the images?

**Comment by kumuji** [comment]
Hi @qiqi-x ,
You can find the place where instances are generated here: https://github.com/kumuji/ugains/blob/82c0c6dee0e5f264316afc154fcbc5fb9e228b9d/ugains/models/sam_sampling.py#L371

And use this function to get boxes from masks: https://github.com/kumuji/ugains/blob/82c0c6dee0e5f264316afc154fcbc5fb9e228b9d/ugains/utils/box_ops.py#L113

Unfortunately, I don't have the code in the repo, since I was using jupyter notebooks, but here is a small snippet that could be useful:
```python
   instance_pred, uncert, inlier_output, nls_uncert = model.predict_step(image)
        for i in range(image.shape[0]):
            boxes = instance_pred[i]["boxes"]
            labels = [f"{j:0.02f}" for j in instance_pred[i]["scores"].flatten()]

            for mask_index in range(instance_pred[i]["masks"].shape[0]):
                img = instance_pred[i]["masks"][mask_index].astype(float)
                colormap_array = virdis_cmap(img)
                img = Image.fromarray((colormap_array[:, :, :3] * 255).astype(np.uint8))
                if save_visualizations:
                    img.save(
                        f"saved/{save_path}/{counter:03d}_{mask_index:02d}_prediction_mask.png"
                    )

            masks = np.concatenate(
                (
                    np.zeros_like(instance_pred[i]["masks"][:1, ...]),
                    instance_pred[i]["masks"],
                ),
                axis=0,
            )
            masks_image = cmap[masks.argmax(0).astype(int)]
            image_with_masks = original_image[i].copy()
            tmp_image_with_masks = overlay_mask_on_image(
                original_image[i], masks_image, mask_opacity=0.6
            )
            msak_img = (masks.argmax(0).astype(int) != 0)[..., None].repeat(3, 2)
            image_with_masks[msak_img] = tmp_image_with_masks[msak_img]

            img = Image.fromarray(original_image[i].astype(np.uint8))
            img.save(
                f"saved/{save_path}/{counter:03d}_original_image.png"
            )

            img = (
                (nls_uncert[i] - nls_uncert[i].min())
                / (nls_uncert[i].max() - nls_uncert[i].min()))
            colormap_array = gnuplot_cmap(img)
            img = Image.fromarray((colormap_array[:, :, :3] * 255).astype(np.uint8))
            img = Image.fromarray(img.numpy().astype(np.uint8))
            img.save(
                f"saved/{save_path}/{counter:03d}_nls_image.png"
            )

            img = (
                (uncert[i] - uncert[i].min())
                / (uncert[i].max() - uncert[i].min()))
            colormap_array = gnuplot_cmap(img)
            img = Image.fromarray((colormap_array[:, :, :3] * 255).astype(np.uint8))
            img.save(
                f"saved/{save_path}/{counter:03d}_uncertainty_image.png"
            )

            inlier_prediction = trainid2color[
                inlier_output.max(1)[1][i].detach().cpu().long().numpy()
            ]
            inlier_prediction = overlay_mask_on_image(
                original_image[i], inlier_prediction, mask_opacity=0.8
            )
            inlier_img = Image.fromarray(inlier_prediction.astype(np.uint8))
            inlier_img.save(
                f"saved/{save_path}/{counter:03d}_inlier_prediction.png"
            )

            boxes = np.concatenate((np.zeros_like(boxes[:1, :]), boxes))
            labels = ["0"] + labels

            image_with_masks = torchvision.utils.draw_bounding_boxes(
                image=torch.from_numpy(image_with_masks.astype(np.uint8)).permute(
                    2, 0, 1
                ),
                boxes=torch.from_numpy(boxes),
                labels=labels,
                width=4,
                font_size=24,
                colors=[tuple(c) for c in cmap.tolist()]
            )
            img_with_masks = Image.fromarray(image_with_masks.permute(1, 2, 0).numpy())
            img_with_masks.save(
                f"saved/{save_path}/{counter:03d}_image_with_masks.png"
            )

            combined_prediction = Image.fromarray(
                overlay_mask_on_image(
                    np.array(image_with_masks).transpose(1, 2, 0),
                    np.array(inlier_img),
                    mask_opacity=0.5,
                )
            )
```

**Comment by qiqi-x** [comment]
Thank you very much for your response! In the first line of the example code you provided, does the model.predict_step() method refer to this method mentioned here? https://github.com/kumuji/ugains/blob/main/ugains/models/sampler.py#L101 However, the return values of this method are different from the ones you gave. In your example code, it returns four values: instance_pred, uncert, inlier_output, and nls_uncert, but the method in this link only returns two values: uncertainty and output. Looking forward to your reply.

**Comment by kumuji** [comment]
It looks more like a `test_step` function from `sam_sampling.py`: https://github.com/kumuji/ugains/blob/82c0c6dee0e5f264316afc154fcbc5fb9e228b9d/ugains/models/sam_sampling.py#L316

`uncertainty` comes from here: https://github.com/kumuji/ugains/blob/82c0c6dee0e5f264316afc154fcbc5fb9e228b9d/ugains/models/sam_sampling.py#L382
`instance_pred` from here: https://github.com/kumuji/ugains/blob/82c0c6dee0e5f264316afc154fcbc5fb9e228b9d/ugains/models/sam_sampling.py#L371
That's `inlier_output`: https://github.com/kumuji/ugains/blob/82c0c6dee0e5f264316afc154fcbc5fb9e228b9d/ugains/models/sam_sampling.py#L328
And `uncert` comes from here: https://github.com/kumuji/ugains/blob/82c0c6dee0e5f264316afc154fcbc5fb9e228b9d/ugains/models/sam_sampling.py#L364

Here is the function I used in my notebook, unedited, but it may raise more questions than it gives answers, please refer to the `test_step` in `sam_sampling.py`.

```python
    def predict_step(self, batch, batch_idx=None, nls_thresh=-0.15, conf_thresh=-0.99):
        x, _ = batch
        query_coordinates, nls, inlier_output = self.test_proposal_sampler(images=x, thresh=nls_thresh)
        ood_output = self.forward(x, query_coordinates, nls)
        
        uncertainty = list()
        instance_prediction = []
        for i in range(len(ood_output["pred_logits"])):
            logits = ood_output["pred_logits"][i].cpu()
            masks = ood_output["pred_masks"][i].cpu()
            boxes_for_instance = batched_mask_to_box(masks)

            if self.nms == "box":
                keep_by_nms = batched_nms(
                    boxes_for_instance.float(),
                    logits,
                    torch.zeros(len(boxes_for_instance)),  # categories
                    iou_threshold=self.nms_thresh,
                )
            elif self.nms == "mask":
                keep_by_nms = self.batched_mask_nms(
                    masks, logits, iou_threshold=self.nms_thresh
                )
                keep_by_nms = torch.Tensor(keep_by_nms).long()
            else:
                keep_by_nms = torch.arange(len(boxes_for_instance))

            if self.post_mask:
                keep_by_ignore = list()
                for j in keep_by_nms.tolist():
                    intersection = torch.sum(masks[j] * self.ignore_mask)
                    if intersection < 100:
                        keep_by_ignore.append(j)
            else:
                keep_by_ignore = keep_by_nms


            # boxes_for_instance = batched_mask_to_box(masks)
            keep_by_ignore = list()
            for j in keep_by_nms.tolist():
                intersection = torch.sum(masks[j] * self.ignore_mask)
                if (intersection < 100) and (logits[j] > conf_thresh):
                    keep_by_ignore.append(j)
            keep_by_nms = torch.arange(len(boxes_for_instance))
            query_coordinates[i] = query_coordinates[i][keep_by_ignore]
            uncertainty.append(
                torch.einsum(
                    "q,qhw->hw",
                    logits[keep_by_ignore].cpu(),
                    masks[keep_by_ignore].float().cpu(),
                )
            )
            instance_prediction.append(
                {
                    "scores": logits[keep_by_ignore].numpy()[..., None],
                    "masks": masks[keep_by_ignore].numpy().astype(np.uint8),
                    "boxes": batched_mask_to_box(masks[keep_by_ignore]).numpy().astype(np.int32),
                }
            )
        uncertainty = torch.stack(uncertainty)
        uncertainty[uncertainty == 0] = -1.0
        nls_uncert = nls.cpu()
        uncertainty = uncertainty + nls_uncert

        return instance_prediction, uncertainty, inlier_output, nls_uncert
```
